### PR TITLE
Add TCP support to Envoy proxy adapter

### PR DESCRIPTION
**Pull request by kimikowang**
Issue #459

### CODE REVIEW

File: sidecar/config/config.go [hunk 1]
```diff
@@ -167,6 +172,8 @@ type Config struct {
 
 	ProxyConfig ProxyConfig `yaml:"proxy_config"`
 
+	TCPProxyConfigs []TCPProxyConfig `yaml:"tcp_proxy_configs"`
+
```
Comment (GregHanson) [inline sidecar/config/config.go]: The `TCPProxyConfigs` should probably be added under the `ProxyConfig` object since it is only used in the proxy code.  And to avoid some repetition in the name it should probably be called just `TCPProxies`

File: sidecar/proxy/envoy/resources.go [hunk 2]
```diff
@@ -108,12 +108,26 @@ type VirtualHost struct {
...
+	DestinationIPList []string `json:"destination_ip_list,omitempty"`
+	DestinationPorts  string   `json:"destination_ports,omitempty"`
+	SourceIPList      []string `json:"source_ip_list,omitempty"`
+	SourcePorts       string   `json:"source_ports,omitempty"`
+}
+
+// HttpRouteConfig definition.
 // See: https://lyft.github.io/envoy/docs/configuration/http_conn_man/route_config/route_config.html#config-http-conn-man-route-table
-type RouteConfig struct {
+type HttpRouteConfig struct {
```
Comment (GregHanson) [inline sidecar/proxy/envoy/resources.go]: name should be HTTPRouteConfig to satisfy `make precommit` conditions

File: sidecar/proxy/envoy/resources.go [hunk 3]
```diff
@@ -129,24 +143,30 @@ type RDS struct {
...
-	CodecType         string       `json:"codec_type"`
-	StatPrefix        string       `json:"stat_prefix"`
-	GenerateRequestID bool         `json:"generate_request_id"`
-	UserAgent         bool         `json:"add_user_agent"`
-	RouteConfig       *RouteConfig `json:"route_config,omitempty"`
-	RDS               *RDS         `json:"rds,omitempty"`
-	Filters           []Filter     `json:"filters"`
-	AccessLog         []AccessLog  `json:"access_log"`
+// HttpFilterConfig definition.
+type HttpFilterConfig struct {
```
Comment (GregHanson) [inline sidecar/proxy/envoy/resources.go]: HTTPFilterConfig

File: testing/apps/tcp/app.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
+#!/usr/bin/env python
+#
+# import socket
+#
+# TCP_IP = '127.0.0.1'
+# TCP_PORT = 1111
+# BUFFER_SIZE = 20  # Normally 1024, but we want fast response
+#
+# s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
```
Comment (GregHanson) [inline testing/apps/tcp/app.py]: remove commented code

File: testing/apps/tcp/test.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
+#!/usr/bin/env python
+#
+# import socket
+#
+#
+# TCP_IP = '127.0.0.1'
+# TCP_PORT = 1111
+# BUFFER_SIZE = 1024
```
Comment (GregHanson) [inline testing/apps/tcp/test.py]: remove commented code

File: testing/apps/tcp/details/details.go [hunk 6]
```diff
@@ -0,0 +1,52 @@
+// Copyright 2016 IBM Corporation
```
Comment (GregHanson) [inline testing/apps/tcp/details/details.go]: is this tcp version of details used in testing at all?

**Review by GregHanson — CHANGES_REQUESTED** [review]
I would recommend adding at least one docker test case to the Travis CI test scenario to make sure our TCP envoy solution works.  A test case for Kubernetes can be added in this PR or as a separate pull request

File: sidecar/proxy/envoy/manager.go [hunk 7]
```diff
@@ -283,6 +246,90 @@ func (m *manager) generateConfig(rules []api.Rule, instances []api.ServiceInstan
 	}, nil
 }
 
+func BuildRDSClusters(tcpProxyList []config.TCPProxyConfig, sdsPort int) []Cluster {
```
Comment (GregHanson) [inline sidecar/proxy/envoy/manager.go]: function should be named `buildStaticClusters` since it is building more than just the `rds` cluster and it shouldn't need to be exported since it is only called in `generateConfig()`

**Review by GregHanson — APPROVED** [review]

File: testing/docker/bookinfo.yaml [hunk 8]
```diff
@@ -33,6 +33,19 @@ gateway:
   container_name: gateway
   command: /usr/bin/run_filebeat.sh
 
+tcphelloworld:
+  image: amalgam8/a8-tcp:latest
+  environment:
+    - A8_REGISTRY_URL=http://registry:8080
+    - A8_SERVICE=tcphelloworld
```
Comment (rshriram) [inline testing/docker/bookinfo.yaml]: Please attach a tag to this service name.
e.g.
`A8_SERVICE=tcphelloworld:version=v1`

File: testing/apps/productpage/config.yaml [hunk 9]
```diff
@@ -4,3 +4,8 @@ commands:
     on_exit: ignore
   - cmd: [ "/opt/microservices/productpage", "9080", "http://localhost:6379" ]
     on_exit: terminate
+
+proxy_config:
+  tcp_proxies:
+    - service: tcphelloworld
```
Comment (rshriram) [inline testing/apps/productpage/config.yaml]: change to use tags, e.g., `tcphelloworld:version=v1`

File: testing/apps/productpage/Dockerfile.sidecar [hunk 10]
```diff
@@ -15,8 +15,8 @@
 FROM amalgam8/a8-sidecar:latest
 
 WORKDIR /opt/microservices
+COPY config.yaml /opt/microservices/
 COPY productpage /opt/microservices/
-COPY config.yaml /opt/microservices/config.yaml
-EXPOSE 9080
+EXPOSE 9080 12345
```
Comment (rshriram) [inline testing/apps/productpage/Dockerfile.sidecar:20]: why should productpage expose 12345 ? Its only for productpage app to Envoy. Nothing is going to talk to productpage over 12345

File: testing/apps/tcp/Makefile [hunk 11]
```diff
@@ -0,0 +1,18 @@
+APP_NAME 	:= $(shell basename `pwd`)
+IMAGE_NAME 	:= amalgam8/a8-$(APP_NAME)
```
Comment (rshriram) [inline testing/apps/tcp/Makefile]: If both client and server are in python, why do we need go based build in teh makefile?
All you need is the dockerfiles and a makefile to create those. i.e. you don't need build targets or build flags or any go related stuff.

File: testing/docker/bookinfo.yaml [hunk 12]
```diff
@@ -115,6 +128,8 @@ productpage-v1:
     - A8_REGISTRY_POLL=5s
     - A8_HEALTHCHECKS=http://localhost:9080/health,file:///bin/ls
     - A8_ELASTICSEARCH_SERVER=elasticsearch:9200
+  ports:
+    - "12345:12345"
```
Comment (rshriram) [inline testing/docker/bookinfo.yaml:132]: why is this needed?

File: sidecar/proxy/envoy/manager.go [hunk 13]
```diff
@@ -196,59 +198,47 @@ func (m *manager) generateConfig(rules []api.Rule, instances []api.ServiceInstan
...
+				},
+			},
+			MaxRequestsPerConnection: 1,
+		},
+	}
+
+	for _, proxy := range m.tcpProxyConfigs {
+		cluster := Cluster{
+			Name:             proxy.Service,
+			ServiceName:      proxy.Service,
```
Comment (rshriram) [inline sidecar/proxy/envoy/manager.go:223]: Why do you need to build clusters for TCP? CDS is already going to be pulling in all the clusters from the service registry. All you need is to build listeners.

Comment (rshriram) [reply sidecar/proxy/envoy/manager.go:223]: The reason I ask is because we have circuit breaker settings. And we need to be able to configure circuit breakers for these services.

**Review by rshriram — CHANGES_REQUESTED** [review]

File: sidecar/proxy/envoy/manager.go (hunk 13 above)
Comment (GregHanson) [reply sidecar/proxy/envoy/manager.go:223]: TCP filters have a statically defined cluster name.  Envoy rejects the config if there is not a statically defined cluster in cluster_manager.  CDS is not compatible with TCP at this point

File: testing/apps/productpage/Dockerfile.sidecar (hunk 10 above)
Comment (GregHanson) [reply testing/apps/productpage/Dockerfile.sidecar:20]: 12345 is a hack to test that the envoy proxy inside productpage can reach the tcphelloworld app in tests.  tcp listeners need to be configured via a config.yaml in sidecar.  We didn't want to modify the gateway image, so we modified productpage since it already had a config yaml to accomplish the same thing

**Review by rshriram — APPROVED** [review]
